test/test_foreach.py

```diff
@@ -911,6 +911,78 @@ def test_outplace_with_invalid_grads(self, device, dtype, op):
         self.assertIsNotNone(sample.input[0].grad)
         self.assertIsNone(sample.input[1].grad)
 
+    @ops(
+        foreach_unary_op_db + foreach_binary_op_db + foreach_pointwise_op_db + foreach_lerp_op_db,
+        dtypes=OpDTypes.supported,
+        allowed_dtypes=(torch.float64, torch.complex128),
+    )
+    def test_outplace_forward_mode_AD(self, device, dtype, op):
+        if not op.supports_forward_ad:
+            self.skipTest("forward AD not supported")
+        if op.name == "_foreach_zero":
+            self.skipTest("`_foreach_zero` not implemented")
+
+        def check_sample_eligibility(op, sample, dtype):
+            if op.name == "_foreach_sub" and (
+                # ScalarList having a bool or Scalar of bool
+                (
+                    isinstance(sample.args[0], list)
+                    and any(isinstance(a, bool) for a in sample.args[0])
+                ) or isinstance(sample.args[0], bool)
+            ):
+                return False, _BOOL_SUB_ERR_MSG
+            rhs_arg_has_complex_number = sample.args and ((
+                isinstance(sample.args[0], list)
+                and any(isinstance(a, complex) for a in sample.args[0])
+            ) or (
+                isinstance(sample.args[0], complex)
+            ))
+            if op.name in ("_foreach_clamp_max", "_foreach_clamp_min") and rhs_arg_has_complex_number:
+                return False, "clamp is not supported for complex types"
+            # Wit this the result will be complex128, so skip.
+            if rhs_arg_has_complex_number and dtype == torch.float64:
+                return False, ""
+
+            return True, ""
+
+        # note(crcrpar): without this, some unary functions fail, unlike inplace and/or complex.
+        value_range = {"low": 0.5, "high": 1.0} if dtype == torch.float64 else {}
+        for sample in op.sample_inputs(
+            device, dtype, requires_grad=True, num_input_tenosrs=[5], same_size=True, **value_range,
+        ):
+            # Skip `_foreach_pow.ScalarAndTensor(Scalar, Tensor[])`
+            if op.name == "_foreach_pow" and isinstance(sample.input, Number):
+                continue
```

Comment:
I still think that we should have a custom sample inputs function for these OpInfos https://github.com/pytorch/pytorch/pull/100695#discussion_r1204662438 (it should be trivial to do now that we have the check_sample_inputs function)

Reply:
would like to repeat what I wrote there. I would get confused so easily with the change

Reply:
Hmm thats fine, its not ideal, but maybe track as a follow up item instead for the future?